src/arrow2/src/compute/zip.rs

```diff
@@ -0,0 +1,174 @@
+        crate::datatypes::DataType::UInt32 => zip_primitive!(crate::array::UInt32Array, u32),
+        crate::datatypes::DataType::UInt64 => zip_primitive!(crate::array::UInt64Array, u64),
+        crate::datatypes::DataType::Float32 => zip_primitive!(crate::array::Float32Array, f32),
+        crate::datatypes::DataType::Float64 => zip_primitive!(crate::array::Float64Array, f64),
+        crate::datatypes::DataType::Date32 => {
+            let mut arr = zip_primitive!(crate::array::Int32Array, i32);
+            arr.change_type(DataType::Date32);
+            arr
+        }
+        crate::datatypes::DataType::Date64 => {
+            let mut arr = zip_primitive!(crate::array::Int64Array, i64);
+            arr.change_type(DataType::Date64);
+            arr
+        }
+        dtype @ crate::datatypes::DataType::Time32(_) => {
+            let mut arr = zip_primitive!(crate::array::Int32Array, i32);
+            arr.change_type(dtype.clone());
+            arr
+        }
+        dtype @ crate::datatypes::DataType::Time64(_) => {
+            let mut arr = zip_primitive!(crate::array::Int64Array, i64);
+            arr.change_type(dtype.clone());
+            arr
+        }
+
+        crate::datatypes::DataType::Binary => {
+            zip!(
+                crate::array::BinaryArray<i32>,
+                crate::array::MutableBinaryArray<i32>
+            )
+        }
+        crate::datatypes::DataType::FixedSizeBinary(size) => {
+            let true_values = true_values
+                .as_any()
+                .downcast_ref::<crate::array::FixedSizeBinaryArray>()
+                .unwrap();
+            let false_values = false_values
+                .as_any()
+                .downcast_ref::<crate::array::FixedSizeBinaryArray>()
+                .unwrap();
+            let mut mutable = <crate::array::MutableFixedSizeBinaryArray>::with_capacity(
+                *size,
+                true_values.len(),
+            );
+            mask.iter().enumerate().for_each(|(i, x)| {
+                if x {
+                    let value = true_values.get(i);
+                    mutable.push(value)
+                } else {
+                    let value = false_values.get(i);
+                    mutable.push(value)
+                }
+            });
+            mutable.as_box()
+        }
+        crate::datatypes::DataType::LargeBinary => {
+            zip!(
+                crate::array::BinaryArray<i64>,
+                crate::array::MutableBinaryArray<i64>
+            )
+        }
+        crate::datatypes::DataType::Utf8 => {
+            zip!(
+                crate::array::Utf8Array<i32>,
+                crate::array::MutableUtf8Array<i32>
+            )
+        }
+
+        crate::datatypes::DataType::LargeUtf8 => {
+            zip!(
+                crate::array::Utf8Array<i64>,
+                crate::array::MutableUtf8Array<i64>
+            )
+        }
+
+        dtype => Err(crate::error::Error::nyi(format!(
```

Comment:
are there any other crucial dtypes I'm missing here? 

The `MutableList`is kinda weird and I couldn't figure out how to get it to work.

Reply:
The growable impl we have above should handle all the dypes.